src/main.lua

```diff
@@ -1667,79 +1681,84 @@ end)
 	local targetCurrentEpochIndex = epochs.getEpochIndexForTimestamp(msg.Timestamp)
 
 	assert(blockHeight, "Block height is required")
 	assert(hashchain, "Hash chain is required")
 
-	msg.ioEvent:addField("Last-Ticked-Epoch-Index", lastTickedEpochIndex)
-	msg.ioEvent:addField("Current-Epoch-Index", lastTickedEpochIndex + 1)
+	msg.ioEvent:addField("Last-Created-Epoch-Index", lastCreatedEpochIndex)
+	msg.ioEvent:addField("Last-Distributed-Epoch-Index", lastDistributedEpochIndex)
 	msg.ioEvent:addField("Target-Current-Epoch-Index", targetCurrentEpochIndex)
 
-	-- if epoch index is -1 then we are before the genesis epoch and we should not tick
-	if targetCurrentEpochIndex < 0 then
-		-- do nothing and just send a notice back to the sender
+	-- if the timestamp is before the genesis epoch start timestamp, do nothing
+	if msg.Timestamp < epochs.getSettings().epochZeroStartTimestamp then
 		Send(msg, {
 			Target = msg.From,
-			Action = "Tick-Notice",
-			LastTickedEpochIndex = LastTickedEpochIndex,
+			Action = "Tick-Notice", -- TODO: this may be better as a "Genesis-Epoch-Not-Started-Notice"
 			Data = json.encode("Genesis epoch has not started yet."),
 		})
 		return
 	end
 
 	-- tick and distribute rewards for every index between the last ticked epoch and the current epoch
-	local tickedEpochIndexes = {}
+	local distributedEpochIndexes = {}
 	local newEpochIndexes = {}
 	local newDemandFactors = {}
-	--- @type PruneGatewaysResult[]
 	local newPruneGatewaysResults = {}
 	local tickedRewardDistributions = {}
 	local totalTickedRewardsDistributed = 0
 
+	print("Ticking from " .. lastCreatedEpochIndex .. " to " .. targetCurrentEpochIndex)
+
 	--- tick to the newest epoch, and stub out any epochs that are not yet created
-	for i = lastTickedEpochIndex + 1, targetCurrentEpochIndex do
-		local _, _, epochDistributionTimestamp = epochs.getEpochTimestampsForIndex(i)
-		-- use the minimum of the msg timestamp or the epoch distribution timestamp, this ensures an epoch gets created for the genesis block
-		-- and that we don't try and distribute before an epoch is created
-		local tickTimestamp = math.min(msg.Timestamp, epochDistributionTimestamp)
-		-- TODO: if we need to "recover" epochs, we can't rely on just the current message hashchain and block height,
-		-- we should set the prescribed observers and names to empty arrays and distribute rewards accordingly
-		local tickResult = tick.tickEpoch(tickTimestamp, blockHeight, hashchain, msgId)
-		if tickTimestamp == epochDistributionTimestamp then
-			-- if we are distributing rewards, we should update the last ticked epoch index to the current epoch index
-			LastTickedEpochIndex = i
-			table.insert(tickedEpochIndexes, i)
-		end
-		Send(msg, {
-			Target = msg.From,
-			Action = "Tick-Notice",
-			LastTickedEpochIndex = tostring(LastTickedEpochIndex),
-			Data = json.encode(tickResult),
-		})
+	for i = lastCreatedEpochIndex, targetCurrentEpochIndex do
```

Comment:
this logically makes more sense - instead of "going back one epoch", start this process at the last known created epoch and go until we reach the target current epoch